modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/TransactionsInMultipleThreads.java

```diff
@@ -0,0 +1,481 @@
+                }
+                catch (Exception e) {
+                    e.printStackTrace();
+                }
+                return null;
+            }
+        });
+    }
+
+    /**
+     * @throws Exception
+     */
+    private void simpleTransactionInAnotherThreadContinued() throws Exception {
+        IgniteCache<String, Integer> cache = jcache(grid(0), cacheConfiguration(null).setName("testCache"), String.class, Integer.class);
+        IgniteTransactions transactions = transactions();
+
+        Transaction tx = transactions.txStart(transactionConcurrency, transactionIsolation);
+
+        cache.put("key1", 1);
+        cache.put("key2", 2);
+        cache.put("key1'", 1);
+
+        tx.stop();
+
+        IgniteInternalFuture<Boolean> fut = GridTestUtils.runAsync(new Callable<Boolean>() {
+            @Override public Boolean call() throws Exception {
+                Assert.assertNull(transactions.tx());
+                Assert.assertEquals(TransactionState.STOPPED, tx.state());
+
+                tx.resume();
+
+                Assert.assertEquals(TransactionState.ACTIVE, tx.state());
+
+                cache.put("key3", 3);
+                cache.put("key2'", 2);
+                cache.remove("key2", 2);
+
+                tx.stop();
+                return true;
+            }
+        });
+        try {
+            fut.get();
+        }
+        catch (IgniteCheckedException e) {
+            fail("transaction in another thread failed.Concurrency control=" + transactionConcurrency +
+                ". Isolation level=" + transactionIsolation);
+        }
+
+        Assert.assertNull(transactions.tx());
+        Assert.assertEquals(TransactionState.STOPPED, tx.state());
+
+        tx.resume();
+
+        Assert.assertEquals(TransactionState.ACTIVE, tx.state());
+
+        cache.remove("key1'", 1);
+        cache.remove("key2'", 2);
+        cache.put("key3'", 3);
+
+        tx.commit();
+
+        Assert.assertEquals(TransactionState.COMMITTED, tx.state());
+        Assert.assertEquals((long)1, (long)cache.get("key1"));
+        Assert.assertEquals((long)3, (long)cache.get("key3"));
+        Assert.assertEquals((long)3, (long)cache.get("key3'"));
+        Assert.assertFalse(cache.containsKey("key2"));
+        Assert.assertFalse(cache.containsKey("key1'"));
+        Assert.assertFalse(cache.containsKey("key2'"));
+
+        cache.removeAll();
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    public void testCrossCacheTransactionInAnotherThread() throws Exception {
+        withAllIsolationsAndConcurrencies(new IgniteClosure<Object, Void>() {
+            @Override public Void apply(Object o) {
+                try {
+                    TransactionsInMultipleThreads.this.crossCacheTransactionInAnotherThread();
+                }
+                catch (Exception e) {
+                    e.printStackTrace();
```

Comment:
May rename "e" -> "ignored"?
If you need to write it to log, please use Logger, otherwise no-operation requered.
New line before "return" required.